File: windows/src/ControlDeck.Domain/DashboardSyncResolver.cs
namespace ControlDeck.Domain;

/// <summary>What a receiving device should do with an incoming DASHBOARD_SYNC.</summary>
public enum SyncOutcome
{
    /// <summary>Apply the incoming dashboard as the new local copy.</summary>
    ApplyIncoming,

    /// <summary>Discard the incoming copy; it is strictly older than local.</summary>
    KeepLocal,

    /// <summary>
    /// Discard the incoming copy AND reply with the local (newer) dashboard so
    /// the sender catches up, per docs/ARCHITECTURE.md §6.
    /// </summary>
    KeepLocalAndReplyWithLocal
}

/// <summary>
/// Implements the last-write-wins rule from docs/ARCHITECTURE.md §6 exactly:
///
/// - No local copy exists yet -> always apply incoming.
/// - incoming.version &gt; local.version -> apply incoming outright.
/// - incoming.version &lt; local.version -> discard, reply with local so the
///   sender catches up.
/// - Equal versions (rare concurrent-edit race) -> tie-break by the enclosing
///   message's `timestamp` (higher wins), then by `sourceDeviceId` string
///   ordering (ordinal / StringComparer.Ordinal) as a final deterministic
///   fallback: the lexicographically GREATER sourceDeviceId wins. Both
///   platforms must implement this identical rule so they converge without
///   further negotiation.
///
/// This is a pure function: no I/O, no mutation. Callers own actually
/// applying the incoming dashboard, sending DASHBOARD_ACK, or re-sending the
/// local DASHBOARD_SYNC.
/// </summary>
public static class DashboardSyncResolver
{
    public static SyncOutcome Resolve(
        bool localDashboardExists,
        long incomingVersion,
        long localVersion,
        long incomingTimestamp,
        long localTimestamp,
        string incomingSourceDeviceId,
        string localSourceDeviceId)
    {
        if (!localDashboardExists)
        {
            return SyncOutcome.ApplyIncoming;
        }

        if (incomingVersion > localVersion)
        {
            return SyncOutcome.ApplyIncoming;
        }

        if (incomingVersion < localVersion)
        {
            return SyncOutcome.KeepLocalAndReplyWithLocal;
        }

        // Equal version: tie-break by timestamp, then by sourceDeviceId ordinal ordering.
        if (incomingTimestamp > localTimestamp)
        {
            return SyncOutcome.ApplyIncoming;
        }

        if (incomingTimestamp < localTimestamp)
        {
            return SyncOutcome.KeepLocalAndReplyWithLocal;
        }

        var comparison = string.CompareOrdinal(incomingSourceDeviceId, localSourceDeviceId);
        return comparison > 0 ? SyncOutcome.ApplyIncoming : SyncOutcome.KeepLocalAndReplyWithLocal;
    }
}

File: android/domain/src/main/kotlin/com/controlldeck/domain/DashboardSyncResolver.kt
package com.controlldeck.domain

/** A DASHBOARD_SYNC as seen by the resolver: the dashboard plus the enclosing envelope metadata. */
data class DashboardSyncMessage(
    val dashboard: Dashboard,
    val timestamp: Long,
    val sourceDeviceId: DeviceId,
)

/** Result of resolving an incoming DASHBOARD_SYNC against local state. */
sealed class DashboardSyncOutcome {
    /** Incoming wins; caller should persist [dashboard] as the new local copy. */
    data class Applied(val dashboard: Dashboard) : DashboardSyncOutcome()

    /** Local is newer/wins the tie-break; caller should send its own DASHBOARD_SYNC back to the sender. */
    data class ReplyWithLocal(val local: Dashboard) : DashboardSyncOutcome()

    /** Same version, identical content — nothing to do. */
    data object NoChange : DashboardSyncOutcome()
}

/**
 * Implements the exact last-write-wins rule from docs/ARCHITECTURE.md §6:
 *
 * 1. Apply incoming iff `incoming.version > local.version` (or no local copy exists).
 * 2. If `incoming.version < local.version`, reply with the local (newer) copy.
 * 3. Equal versions: tie-break by message `timestamp`, then by `sourceDeviceId`
 *    string ordering, deterministically on both sides.
 */
object DashboardSyncResolver {

    fun resolve(local: DashboardSyncMessage?, incoming: DashboardSyncMessage): DashboardSyncOutcome {
        if (local == null) return DashboardSyncOutcome.Applied(incoming.dashboard)

        val localVersion = local.dashboard.version
        val incomingVersion = incoming.dashboard.version

        return when {
            incomingVersion > localVersion -> DashboardSyncOutcome.Applied(incoming.dashboard)
            incomingVersion < localVersion -> DashboardSyncOutcome.ReplyWithLocal(local.dashboard)
            else -> resolveTie(local, incoming)
        }
    }

    private fun resolveTie(local: DashboardSyncMessage, incoming: DashboardSyncMessage): DashboardSyncOutcome {
        if (local.dashboard == incoming.dashboard) return DashboardSyncOutcome.NoChange

        return when {
            incoming.timestamp > local.timestamp -> DashboardSyncOutcome.Applied(incoming.dashboard)
            incoming.timestamp < local.timestamp -> DashboardSyncOutcome.ReplyWithLocal(local.dashboard)
            else -> {
                // Final deterministic fallback: sourceDeviceId string ordering. The
                // lexicographically greater id "wins" — arbitrary but symmetric, so
                // both peers independently compute the same winner.
                val incomingWins = incoming.sourceDeviceId.value > local.sourceDeviceId.value
                if (incomingWins) {
                    DashboardSyncOutcome.Applied(incoming.dashboard)
                } else {
                    DashboardSyncOutcome.ReplyWithLocal(local.dashboard)
                }
            }
        }
    }
}
